### PR TITLE
Langchain integration

### PR DESCRIPTION
**Pull request by hrobbins**
Trying to keep things simple for this initial conversion to LangChain.  The only difference with the original functionality is this version doesn't stream the output from OpenAI, so there is bit of a delay in waiting for the result.

This branch will enable using any LLM, from hugging-face, locally, any LangChain supported LLM service, such as Anthropic/Claude, or pointing it to a Custom LLM.

### CODE REVIEW

File: gpt_engineer/ai.py [hunk 1]
```diff
@@ -1,57 +1,88 @@
...
-            chat.append(msg)
-        print()
-        messages += [{"role": "assistant", "content": "".join(chat)}]
+        r = self.chat(messages)
+        messages += [r]  # AI Message
+
         logger.debug(f"Chat completion finished: {messages}")
         return messages
 
+    def lastMessageContent(self, messages):
```
Comment (AntonOsika) [inline gpt_engineer/ai.py:60]: python uses snake case

File: gpt_engineer/ai.py [hunk 2]
```diff
@@ -1,57 +1,88 @@
...
+        return m
+
+    def serializeMessages(messages):
+        # dicts = messages_to_dict(history.messages)
+        r = "[]"
+        try:
+            if messages and isinstance(messages, list) and len(messages) > 0:
+                r = json.dumps(messages_to_dict(messages))
+        except Exception as e:
+            logging.warn("Exception serializing messages, returning empty array", e)
```
Comment (AntonOsika) [inline gpt_engineer/ai.py:75]: This exception handling should not exist imo?

File: gpt_engineer/learning.py [hunk 3]
```diff
@@ -97,15 +98,19 @@ def logs_to_string(steps: List[Step], logs: DB):
     chunks = []
     for step in steps:
         chunks.append(f"--- {step.__name__} ---\n")
-        messages = json.loads(logs[step.__name__])
-        chunks.append(format_messages(messages))
+        # messages = json.loads(logs[step.__name__])
```
Comment (AntonOsika) [inline gpt_engineer/learning.py:101]: comments?

**Review by AntonOsika — CHANGES_REQUESTED** [review]
Not 100% we want langchain in here:)

**Comment by AntonOsika** [comment]
Thanks @hrobbins for contributing to improving the repository!

I can see the case for being able to use other language models, but I think we can do that without langchain, and we want to keep it simple. Are there other advantages?

We only merge PRs that use python pep8, doesn't leave comments hanging around etc. So since this is not the focus right now, I will close the PR, but feel free to open it if you are committed to fixing everything and can argue for why we want langchain **today**.

File: gpt_engineer/ai.py [hunk 4]
```diff
@@ -1,57 +1,88 @@
...
+    def __init__(self, modelid="gpt-4", temperature=0.1):
         self.temperature = temperature
-        self.model = model
+        self.modelid = modelid
+        # llm = load_llm("llm.yaml")  # load llm from file
+        try:
+            self.chat = ChatOpenAI(model=self.modelid, temperature=temperature)
+        except Exception as e:
+            # ? try a different model
+            logging.warning(e)
```
Comment (AntonOsika) [inline gpt_engineer/ai.py:29]: Should not catch this error – it should raise!

File: gpt_engineer/ai.py [hunk 5]
```diff
@@ -1,57 +1,88 @@
...
-        messages += [{"role": "assistant", "content": "".join(chat)}]
+        r = self.chat(messages)
+        messages += [r]  # AI Message
+
         logger.debug(f"Chat completion finished: {messages}")
         return messages
 
+    def lastMessageContent(self, messages):
+        m = messages[-1].content
+        if m:
```
Comment (AntonOsika) [inline gpt_engineer/ai.py:62]: Doesn't need if statement here

File: gpt_engineer/ai.py [hunk 6]
```diff
@@ -1,57 +1,88 @@
...
+            m = m.strip()
+        # logging.info(m)
+        print(m)
+        return m
+
+    def serializeMessages(messages):
+        # dicts = messages_to_dict(history.messages)
```
Comment (AntonOsika) [inline gpt_engineer/ai.py:69]: ?

File: gpt_engineer/ai.py [hunk 7]
```diff
@@ -1,57 +1,88 @@
...
+        r = "[]"
+        try:
+            if messages and isinstance(messages, list) and len(messages) > 0:
```
Comment (AntonOsika) [inline gpt_engineer/ai.py:72]: Why check?

**Comment by AntonOsika** [comment]
There is a lot of good stuff here so happy to reopen

**Comment by AntonOsika** [comment]
@hrobbins thought more about it and I like many things here

would need to address the comments and make sure streaming is still working before we merge though. 

Have you given up on this or are you up for the challenge?

**Comment by UmerHA** [comment]
@hrobbins if you don't have the time to do it, let me know - id be happy to do it!

need this to finish https://github.com/AntonOsika/gpt-engineer/pull/431

(cc @AntonOsika)

**Comment by hrobbins** [comment]
Can probably polish it up today.
Already made most of the recommended changes, looks like it just needs a
callback handler for the streaming.

For other reasons, I started working on breaking down the processes into
more isolated requests to the LLM so each one is individually a smaller
response anyway.  Streaming is less important when we're not asking it to
do everything at once with each call.  But I can go ahead and re-add it.

-H

**Comment by UmerHA** [comment]
Hey @hrobbins! Sorry, I was too eager 🙈 I had already implemented it before I saw your response.

Here's the branch: https://github.com/UmerHA/gpt-engineer/commits/langchain-integration.

Iiuc, the scope of this PR is to make it work for (i) LangChain models, (ii) HuggingFace models, and (iii) local models. I've done (i), so you could still do (ii) and (iii).

If you want, I can merge my code into yours?
Or should I open a PR instead?

Will do what you decide.

**Comment by hrobbins** [comment]
No worries, go for it.

-H

**Comment by hrobbins** [comment]
> Hey @hrobbins! Sorry, I was too eager 🙈 I had already implemented it before I saw your response.
> 
> Here's the branch: https://github.com/UmerHA/gpt-engineer/commits/langchain-integration.
> 
> Iiuc, the scope of this PR is to make it work for (i) LangChain models, (ii) HuggingFace models, and (iii) local models. I've done (i), so you could still do (ii) and (iii).
> 
> If you want, I can merge my code into yours? Or should I open a PR instead?
> 
> Will do what you decide.

Reviewed what you did.  Nicely done!

**Pull request closed without merging**